src/FS.Embedded.Manifest.Task/Entry.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Extensions.FileProviders.Embedded.Manifest.Task
+{
+    [DebuggerDisplay("{Name,nq}")]
+    public class Entry : IEquatable<Entry>
+    {
+        public bool IsFile { get; private set; }
+
+        public string Name { get; private set; }
+
+        public string AssemblyResourceName { get; private set; }
+
+        public ISet<Entry> Children { get; } = new SortedSet<Entry>(NameComparer.Instance);
+
+        public static Entry Directory(string name) =>
+            new Entry { Name = name };
+
+        public static Entry File(string name, string assemblyResourceName) =>
+            new Entry { Name = name, AssemblyResourceName = assemblyResourceName, IsFile = true };
+
+        public void AddChild(Entry child)
+        {
+            if (IsFile)
+            {
+                throw new InvalidOperationException("Tried to add children to a file.");
+            }
+
+            var existing = Children.SingleOrDefault(c => c.HasName(child.Name));
+            if (existing != null && !existing.Equals(child))
+            {
+                throw new InvalidOperationException("Tried to add two different items with the same name");
+            }
+
+            Children.Add(child);
```

Comment:
You don't have a check for a situation, where the child has already been added, I would throw in that case also, instead of silently ignoring the request